### PR TITLE
Feature progress speed

### PR DESCRIPTION
**Pull request by MstrPikachu**
Now we can calculate speed in the last x seconds

### CODE REVIEW

**Comment by MstrPikachu** [comment]
Behaviour between both calculators are slightly different, the fixed window tends to not be as smooth in test case 3. idk why

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java [hunk 1]
```diff
@@ -276,10 +277,10 @@ public String getVersion(String name) {
         return "";
     }
 
-    /**
-     * Compares the hash of the locally-installed plugin of a specified 
-     * version with the hash of the plugin of that version on the server
-     * @return True if the hashes match, false otherwise
+    /*
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java]: Is this the right format for docstrings? It probably doesn't make much of a difference, but just making sure.

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java [hunk 2]
```diff
@@ -136,6 +136,7 @@ public void addManuallyInstalled(String name)
                 } catch (IOException e) {
                     System.out.printf("Error updating CSV");
                 }
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java:138]: Deprecated in favor of new setManuallyInstalled method, no point in keeping this.

File: src/main/java/org/hydev/mcpm/client/database/PluginTracker.java [hunk 3]
```diff
@@ -10,7 +10,7 @@
 import java.util.ArrayList;
 import java.io.*;  
 
-interface PluginTracker {
+public interface PluginTracker {
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/database/PluginTracker.java:13]: Good catch

File: src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByName.java [hunk 4]
```diff
@@ -26,7 +26,7 @@ public Map<String, List<PluginModel>> constructSearchMaps(List<PluginModel> plug
         for (PluginModel plugin : plugins) {
             // Get latest version
             var v = plugin.getLatestPluginVersion();
-            if (v.isPresent() && v.get().meta() != null && !v.get().meta().name().equals("")) {
+            if (v.isPresent() && v.get().meta() != null && v.get().meta().name() != null && !v.get().meta().name().equals("")) {
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByName.java:29]: Line may be too long.

File: src/main/java/org/hydev/mcpm/client/interaction/ProgressBar.java [hunk 5]
```diff
@@ -81,12 +81,12 @@ public ProgressRowBoundary appendBar(ProgressRowBoundary bar)
     }
 
 
-    public void incrementBarProgress(int id, long inc) {
+    private void incrementBarProgress(int id, long inc) {
         this.bars.get(id).increase(inc);
     }
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/interaction/ProgressBar.java:87]: Missing docstring

File: src/main/java/org/hydev/mcpm/client/interaction/ProgressBar.java [hunk 6]
```diff
@@ -81,12 +81,12 @@ public ProgressRowBoundary appendBar(ProgressRowBoundary bar)
...
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/interaction/ProgressBar.java:83]: Missing docstring

File: src/main/java/org/hydev/mcpm/client/interaction/ProgressSpeedCalculator.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
+package org.hydev.mcpm.client.interaction;
+
+import org.hydev.mcpm.utils.Pair;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * An implementation of ProgressSpeedCalculator using a queue and sliding window method.
+ *
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/interaction/ProgressSpeedCalculator.java:10]: May be beneficial to include a more substantial docstring explaining what the ProgressSpeedCalculator actually does lol

File: src/main/java/org/hydev/mcpm/client/interaction/VariableWindowProgressSpeedCalculator.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    // then psa will not contain 0 for this instance of time
+    // in other words we use coordinate compression
+
+    private final TreeMap<Long, Integer> map = new TreeMap<>(); // maps time in nanoseconds to psa indices
+    private long total = 0;
+    private long start = 0; // start time
+    private final long defaultWindow;
+
+    /**
+     * Constructs a VariableWindowProgressSpeedCalculator with the given default window for getSpeed in nanoseconds.
```
Comment (kchprog) [inline src/main/java/org/hydev/mcpm/client/interaction/VariableWindowProgressSpeedCalculator.java:28]: Is such precision actually necessary? Else, things look good.

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java (hunk 1 above)
Comment (MstrPikachu) [reply src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java]: This isn’t my code idk why it’s here how do I sync it with main

**Review by kchprog — CHANGES_REQUESTED** [review]

File: src/main/java/org/hydev/mcpm/client/interaction/VariableWindowProgressSpeedCalculator.java (hunk 8 above)
Comment (MstrPikachu) [reply src/main/java/org/hydev/mcpm/client/interaction/VariableWindowProgressSpeedCalculator.java:28]: I just want it to be consistent with the other one

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java (hunk 1 above)
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java]: It's not the right format for docstrings, but the function below is commented out so having a docstring on nothing gives checkstyle errors so Peter changed it. It's fine

File: src/main/java/org/hydev/mcpm/client/database/export/Export.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
+package org.hydev.mcpm.client.database.export;
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/export/Export.java:1]: AHHH export is just spreading across pre this is creeping me out.

**Review by 1whatleytay — APPROVED** [review]

**Comment by 1whatleytay** [comment]
Merge this!!

**Review by hykilpikonna — APPROVED** [review]

**Pull request merged**